src/cheri_insts.sail

```diff
@@ -733,31 +733,25 @@ function clause execute (CBuildCap(cd, cs1, cs2)) = {
 
 union clause ast = CCopyType : (regidx, regidx, regidx)
 /*!
- * If the **otype** field of capability register *cs2* is not a reserved value
- * then capability register *cd* is replaced with the contents of capability
- * register *cs1* with the **address** set to *cs2*.**otype**. Otherwise
- * capability register *cd* is replaced with the **NULL** capability with its
- * **address** set to *cs2*.**otype**.
+ * Capability register *cd* is replaced with the contents of capability
+ * register *cs1* with the **address** set to *cs2*.**otype** and the
+ * **tag** field cleared if *cs2* has a reserved **otype**.
  *
  * ## Notes
  *
- * - Reserved otypes always get turned into derived-from-**NULL** capabilities,
- *   regardless of the authority given in *cs1*, as, at the moment, all
- *   reserved otypes are constructed using ambiently-available actions.
- *   [CCSeal] knows how to work with these.
+ * - Reserved otypes always result in untagged capabilities, as, at the
+ *   moment, all reserved otypes are constructed using ambiently-available
+ *   actions. [CCSeal] knows how to work with these.
  */
 function clause execute (CCopyType(cd, cs1, cs2)) = {
   let cs1_val = clearTagIfSealed(C(cs1));
   let cs2_val = C(cs2);
-  let cs2_otype = unsigned(cs2_val.otype);
-  if hasReservedOType(cs2_val) then {
-    C(cd) = int_to_cap(EXTS(cs2_val.otype));
-    RETIRE_SUCCESS
-  } else {
-    let (representable, newCap) = setCapAddr(cs1_val, to_bits(cap_addr_width, cs2_otype));
-    C(cd) = clearTagIf(newCap, not(representable));
-    RETIRE_SUCCESS
-  }
+  let reserved = hasReservedOType(cs2_val);
+  let otype : xlenbits = if reserved then EXTS(cs2_val.otype)
+                                     else EXTZ(cs2_val.otype);
```

Comment:
Oh, that makes me kind of nervous... at least a big comment about all reserved otypes being "negative" right now?

Reply:
In what sense? You'll find the same code in CGetType.